posts/2024-09-05_writing_my_first.../writing_my_first_custom_ci_cd_action_for_the_pharmaverseblog.qmd

```diff
@@ -0,0 +1,187 @@
+  names(error_messages) <- rep("x", length(error_messages) - 1)
+
+  concatenated_error_messages <- cli::cli_bullets(error_messages)
+
+  cli::cli_abort(concatenated_error_messages)
+}
+```
+
+The script works as follows:
+
+1. Get a full list of blog posts. These are all the `.qmd` files within the `posts` folder of the pharmaverseblog repo.
+2. Specify a vector of "allowed tags" in the `allowed_tags.R` script.
+3. Specify a function that, given a post:
+    a. Extracts categories from the `yaml` header of the `.qmd` file.
+    b. Cross-checks the tags with the allowed ones.
+    c. For problematic tags, uses the `{cli}` package to construct a nicely-formatted error message.
+4. Loop `check_post_tags()` over all blog posts using a simple `lapply()` call and extract all error messages.
+5. If there are any error messages, use `{cli}` again to construct a concatenated error message.
+
+The final error message looks something like the below:
+
+```
+✖ The tag(s) "ADaMs" in the post
+  "2024-06-17_new_admiral_ex.../new_admiral_extension_packages_admiralpeds_coming_soon.qmd"
+  are not from the allowed list of tags.
+Please select from the following tags: "Metadata", "SDTM", "ADaM", "TLG", "Shiny",
+"Community", "Conferences", "Submissions", and "Technical", or contact one of the maintainers.
+```
+
+## Creating a pipeline for the check
+
+When it came to creating a pipeline for the check, if you had asked me to do this a few months ago, I wouldn't have known my left from my right. Luckily I had recently attended a great CICD workshop at useR 2024 in Salzburg, led by Daphne Grasselly, Franciszek Walkowiak and Pawel Rucki. You can find the repository from their workshop [here](https://github.com/user-workshop-cicd/r.package.example) - it was invaluable to orient me in the right direction. With a (very naive) google search, I also found [this video](https://www.youtube.com/watch?v=u-TcADhWosA), which shows how to execute an R script automatically whenever a pull request is made to the `main` branch of a repo.
+
+Through some trial and error, I was able to coalesce the above resources into quite a short `yaml` file that set up my CICD pipeline. Within the pharmaverseblog repository, the CICD pipelines live under `.github/workflows`. There, I added the following new workflow, in the form of a `yaml` file, called `check_post_tags.yml`:
+
+``` 
+name: Check post tags
+
+on:
+  pull_request:
+    branches:
+      - 'main'
+
+jobs:
+  Check-post-tags:
+    runs-on: ubuntu-latest
+    container:
+      image: "rocker/tidyverse:4.2.1"
+    steps:
+      - uses: actions/checkout@v4
+      - name: Run check_post_tags
+        run: source("R/check_post_tags.R")
+        shell: Rscript {0}
+```
+
+It's deceptively simple to read:
+
+1. We execute the workflow upon any pull request to main.
+2. When there is a pull request to main, we can use the `rocker/tidyverse:4.2.1` to install all the `{tidyverse}` packages.
```

Comment:
The tidyverse packages come pre-installed in the docker image, no installation going on there I think

Reply:
thanks, updated